examples/albef/model.py

```diff
@@ -0,0 +1,361 @@
+        prediction_head: PredictionHead,
+    ) -> None:
+        super().__init__()
+        self.text_embeddings = text_embeddings
+        self.multimodal_encoder = multimodal_encoder
+        self.prediction_head = prediction_head
+
+    def forward(
+        self,
+        input_ids: Tensor,
+        attention_mask: Tensor,
+        encoder_hidden_states: Tensor,
+        encoder_attention_mask: Tensor,
+    ) -> Tensor:
+        hidden_states = self.text_embeddings(input_ids)
+        attention_mask = get_extended_attention_mask_for_decoder(attention_mask)
+        decoder_output = self.multimodal_encoder(
+            hidden_states=hidden_states,
+            attention_mask=attention_mask,
+            encoder_hidden_states=encoder_hidden_states,
+            encoder_attention_mask=encoder_attention_mask,
+        )
+        prediction_scores = self.prediction_head(decoder_output)
+        return prediction_scores
+
+
+class ALBEFModelForVQA(nn.Module):
+    """
+    ALBEF Model for VQA finetuning and inference.
+
+    Args:
+        model (ALBEFModel): Instantiated ALBEFModel.
+        answer_decoder (ALBEFDecoder): Instantiated ALBEFDecoder.
+        loss (MaskedLanguageModelingLoss): Instantiated MaskedLanguageModelingLoss.
+
+    Inputs:
+        image (Tensor of shape (B, C, H, W)): Image features.
+        question (Tensor of shape (B, L)): Question text features.
+        question_atts (Tensor of shape (B, L)): Question attention mask.
+        answers (Tensor of shape (N, M)): Answer text features.
+        answers_atts (Tensor of shape (N, M)): Answer attention mask.
+        ans_weights (Optional[Tensor] of shape (N)): Weights for each answer.
+            Required if is_train is True.
+        ans_lengths (Optional[List[int]] of length B): Number of answers for each question.
+            ans_lengths should sum to N.
+            Required if is_train is True.
+        alpha (Optional[float]): The interpolation value between mlm_loss and loss_distill.
+            Required if is_train is True.
+        k (Optional[int]): The number of answers to return for inference.
+            Required if is_train is False.
+        is_train (Optional[bool]): Whether the model is in training.
+
+    Returns:
+        is_train is True:
+            Tensor: The masked language loss for input.
+        is_train is False:
+            Tuple[Tensor, Tensor]: The ids and probabilities for the top k predicted answers.
+    """
+
+    def __init__(
+        self,
+        model: ALBEFModel,
+        answer_decoder: ALBEFDecoder,
+        loss: MaskedLanguageModelingLoss,
+    ) -> None:
+        super().__init__()
+        self.model = model
+        self.answer_decoder = answer_decoder
+        self.loss = loss
+        self.answer_decoder_m = copy.deepcopy(self.answer_decoder)
+        remove_grad(
+            self.answer_decoder_m
+        )  # remove gradient for the momentum decoder model
+
+    def _train_forward(
```

Comment:
Both these forwards are hard to follow (I think this is unavoidable). But maybe a brief docstring for each method explaining at a high level the steps it is taking would help